### DISCUSSION TITLE
Lamella partition help!

### DISCUSSION THREAD

**Discussion by Dreamown** [neperfepx/neper#606] [Q&A] [answered]
I use python to develop a lamella partition software. This graph comes from that software. In each voronoicell, lamella of different thickness can be assigned to a different set in the software. However, the later-stage meshing presented great difficulties, and I am not an expert in meshing. To my surprise, neper also has the ability to lamella partition, but I have yet to figure out how to implement the collection of cells I described above like in my software.
Just like,in neper i can use below command to generate multiscale partition:
Neper -T -n 10::from_morpho -morpho ''gg::lamella(w=0.01:0.03)'' -o test

How do I go on? I hope you can help me
![image](https://user-images.githubusercontent.com/32787082/201976157-d1a52277-48e6-47f2-85f0-007bd75429eb.png)

**Comment by rquey** [comment, marked as answer]
Yes, Neper does generate (and meshes) multi-scale / lamellar structures.  See these papers:
- meshing: https://neper.info/papers.html#id1
- applications of lamellar structures: https://www.sciencedirect.com/science/article/pii/S0022509616306962 https://www.sciencedirect.com/science/article/pii/S0022509616306962

Example:
```
neper -T -n 10::from_morpho -morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001
neper -M test.tess -pl 100
neper -V test.tess,test.msh -showelt1d all -dataelt1drad 0.003 -dataelt3dedgerad 0.001 -dataelsetcol "scaleid(1)" -print img1 -dataelsetcol id -print img2
```
![image](https://user-images.githubusercontent.com/2659376/202121646-efb38e95-0f78-4a3a-b69e-61286aa59885.png)
![image](https://user-images.githubusercontent.com/2659376/202121658-1ae30afa-4327-4874-8db8-8928a96d126f.png)

**Comment by Ali141272** [comment]
Hello @rquey,
Professor when I run the same commands on Ubuntu I have face some error (BFGS Update error), I don't know why?
My Neper version is 4.6.0.
I used the following commands:

neper -T -n 10::from_morpho -morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001
neper -M test.tess -pl 100
neper -V test.tess,test.msh -showelt1d all -dataelt1drad 0.003 -dataelt3dedgerad 0.001 -dataelsetcol "scaleid(1)" -print img1 -dataelsetcol id -print img2

Thank you.

<img width="760" alt="AE06E575-B63F-4dca-AB9F-283983011DCD" src="https://github.com/neperfepx/neper/assets/102291665/a32b903a-a11c-4ca4-81a8-b6a8d0d733fa">

**Reply by rquey** [reply]
You can simply disregard the "BFGS" errors (they come from Gmsh).

**Reply by Ali141272** [reply]
Thank Professor Romain Quey  for your valuable suggestions.
I have some quarriers regarding two dimension:
1) When I convert the above commands into 2D, then its shows error called (core dumped) as shown in picture using the following command;
neper -T -n 20::from_morpho -morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001 -morpho "diameq:lognormal(1,0.05),1-sphericity:lognormal(0.145,0.03)" -domain "square(0.05,0.1)" -dim 2  
2) Should I change the grain size and grain shape using -morpho "diameq:lognormal(1,0.05),1-sphericity:lognormal(0.145,0.03)" as mentioned in query (1)?
3) For meshing should I used neper -M test.tess -pl 100, but if I will change the element size and mesh quality should I will use -rcl 1.5 -elt tri -meshqualmin 1.0 in neper -M test.tess -pl 100.?

Thank you.

<img width="1120" alt="2A2ABB9B-276C-434e-B0DB-8676FD9A3068" src="https://github.com/neperfepx/neper/assets/102291665/81e1c03f-c2a7-40e2-a84b-988d91f4024e">

**Reply by rquey** [reply]
Please always copy the terminal output as text (not as an image).

Your command contains 2 `-morpho` instances. Remove one and make `-n` consistent with it.

**Reply by Ali141272** [reply]
sajid@DESKTOP-UTRLKKU:/mnt/c/Ubuntu_softwares$ export OMP_NUM_THREADS=1
sajid@DESKTOP-UTRLKKU:/mnt/c/Ubuntu_softwares$ neper -T -n 20::from_morpho -morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001 -morpho "diameq:lognormal(1,0.05),1-sphericity:lognormal(0.145,0.03)" -domain "square(0.05,0.1)" -d
im 2

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.6.0
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 1 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/sajid/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] -n 20::from_morpho -morpho gg::lamellar(w=0.01:0.03)
         -o test -reg 1 -rsel 0.001 -morpho
         diameq:lognormal(1,0.05),1-sphericity:lognormal(0.145,0.03)
         -domain square(0.05,0.1) -dim 2
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellations...
Info   :   [level] Running in multiscale mode (2 scales)...
Info   :   [    1]  1/1 Info   :   - Setting seeds... 100%
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =7.668179813
Info   :     > Iteration   2306: fmin=0.000262055 f=0.000262075
Info   :     > Final solution  : f   =0.000262055 (2306 iterations)
Info   :     > Reached `eps' criterion.
Info   :   [    2]  1/20 Error  :       . Could not process '-n from_morpho'.
Aborted (core dumped)

**Reply by Ali141272** [reply]
1) neper -T -n 20::from_morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001 -morpho "diameq:lognormal(1,0.05),1-sphericity:lognormal(0.145,0.03)" -domain "square(0.05,0.1)" -dim 2
Here I remove one -morpho and still show error:
Error  : Bad arguments!
Aborted (core dumped)

2)  neper -T -n 20::from_morpho -morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001 "diameq:lognormal(1,0.05),1-sphericity:lognormal(0.145,0.03)" -domain "square(0.05,0.1)" -dim 2

>>> Here I removed -morpho from the other side but in both cases not run and show error like;
Error  : Bad arguments!
Aborted (core dumped)

**Reply by Ali141272** [reply]
neper -T -n 20::from_morpho -morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001 -morpho "diameq:lognormal(1,0.05),1-sphericity:lognormal(0.145,0.03)" -domain "square(0.05,0.1)" -dim 2
Professor, this is the complete command you mentioned that you have used 2 -morpho instances.

**Reply by Ali141272** [reply]
Hllo @rquey  I have run the commands:

neper -T -n 100:: -morpho "gg::lamellar(w=0.01:0.03)" -o test -reg 1 -rsel 0.001 -morpho "diameq:lognormal(2,0.04),1-sphericity:lognormal(0.055,0.03)" -domain "square(0.05,0.1)" -dim 2
neper -M test.tess -pl 100
neper -V test.tess,test.msh -showelt1d all -dataelt1drad 0.000001 -dataelt3dedgerad 0.001 -dataelsetcol "scaleid(1)" -print img1 -dataelsetcol id -print img5

But the png flile is totally black and when I open the test.msh file in Gmsh then there is no laminations.

Thank you.
<img width="666" alt="AA3D7A0A-7AB6-4d63-813B-BB2B4ED18050" src="https://github.com/neperfepx/neper/assets/102291665/3323b43e-42c1-45e3-a724-7ac76e7f26d6">
